File: js/chess-notation/src/styles/apply.ts
import { Check, Mate, Castle, PawnCapture, PieceCapture } from './styles.js';
import { isCastle, isCheck, isMate, isPawnCapture, isPieceCapture } from './classify.js';
import { castle, check, mate, pawnCapture, pieceCapture, CapturedPiece } from './map.js';

export type KingEffects = {
  check?: Check;
  mate?: Mate;
}

export type Motions = {
  castle?: Castle;
  pawnCapture?: PawnCapture;
  pieceCapture?: PieceCapture;
}

export type Styles = KingEffects & Motions

export function styleMotions(san: string, motions: Motions, capturedPiece?: CapturedPiece): string {
  if (isPawnCapture(san)) {
    return pawnCapture(san, motions.pawnCapture ?? 'dxe4', capturedPiece);
  } else if (isPieceCapture(san)) {
    return pieceCapture(san, motions.pieceCapture ?? 'Nxe4', capturedPiece);
  } else if (isCastle(san)) {
    return castle(san, motions.castle ?? 'O-O');
  } else {
    return san;
  }
}

export function styleEffects(san: string, effects: KingEffects): string {
  if (isCheck(san)) {
    return check(san, effects.check ?? 'NONE');
  } else if (isMate(san)) {
    return mate(san, effects.mate ?? 'NONE');
  } else {
    return san;
  }
}

export function style(san: string, styles: Styles, capturedPiece?: CapturedPiece): string {
  return styleEffects(styleMotions(san, styles, capturedPiece), styles);
}

File: python/chess-notation/src/chess_notation/styles/apply.py
from .styles import Motions, KingEffects, Styles, no_na
from .classify import is_castle, is_check, is_mate, is_pawn_capture, is_piece_capture
from .map import castle, check, mate, pawn_capture, piece_capture, CapturedPiece

def style_motions(san: str, motions: Motions, captured_piece: CapturedPiece | None = None) -> str:
  if is_pawn_capture(san):
    return pawn_capture(san, no_na(motions.pawn_capture) or 'dxe4', captured_piece)
  elif is_piece_capture(san):
    return piece_capture(san, no_na(motions.piece_capture) or 'Nxe4', captured_piece)
  elif is_castle(san):
    return castle(san, no_na(motions.castle) or 'O-O')
  else:
    return san
  
def style_effects(san: str, effects: KingEffects) -> str:
  if is_check(san):
    return check(san, no_na(effects.check) or 'NONE')
  elif is_mate(san):
    return mate(san, no_na(effects.mate) or 'NONE')
  else:
    return san
  
def style(san: str, styles: Styles, captured_piece: CapturedPiece | None = None) -> str:
  return style_effects(style_motions(san, styles, captured_piece), styles)
